In [1]:
import pandas as pd

boston = pd.read_csv("boston.csv")

boston.head()

CRIM    ZN  INDUS  CHAS     NX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

In [2]:
boston.tail()

CRIM   ZN  INDUS  CHAS     NX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
501  0.06263  0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273     21.0   
502  0.04527  0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273     21.0   
503  0.06076  0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273     21.0   
504  0.10959  0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273     21.0   
505  0.04741  0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273     21.0   

          B  LSTAT  MEDV  
501  391.99   9.67  22.4  
502  396.90   9.08  20.6  
503  396.90   5.64  23.9  
504  393.45   6.48  22.0  
505  396.90   7.88  11.9

In [3]:
boston.shape

(506, 14)

In [4]:
boston.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NX       506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    int64  
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(11), int64(3)
memory usage: 55.5 KB


In [5]:
x = boston.drop("MEDV", axis=1).values #input factors

y = boston["MEDV"].values
#values method converts a datafarme into array

In [6]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()#instantiation

xs = scaler.fit_transform(x) #result is in form of array

In [7]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(xs, y, test_size=0.3,\
                                                    random_state=1)

In [8]:
x_train.shape

(354, 13)

In [9]:
x_test.shape

(152, 13)

In [10]:
from sklearn.linear_model import LinearRegression

#instantiate regression
model = LinearRegression()

model.fit(x_train,y_train)

LinearRegression()

y = a + bx, simple linear regression

y = a + bx1 +cx2 +dx3+ .......

y = 5 + 1.5 x

In [11]:
model.intercept_

22.589670302295588

In [12]:
model.coef_

array([-0.84677953,  1.41623312,  0.40553633,  0.61902108, -2.48543003,
        1.96244469,  0.10052138, -3.18967316,  2.67519835, -1.89922198,
       -2.17462796,  0.58828654, -4.05806653])

In [13]:
o = boston.drop("MEDV", axis=1)
o.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT'],
      dtype='object')

In [14]:
pd.DataFrame(model.coef_,index=o.columns)

0
CRIM    -0.846780
ZN       1.416233
INDUS    0.405536
CHAS     0.619021
NX      -2.485430
RM       1.962445
AGE      0.100521
DIS     -3.189673
RAD      2.675198
TAX     -1.899222
PTRATIO -2.174628
B        0.588287
LSTAT   -4.058067

In [15]:
#r squared:  coeff of determination

model.score(x_test,y_test)

0.7836295385076292

In [16]:
#margin of error: RMSE

y_pred = model.predict(x_test)

from sklearn.metrics import mean_squared_error
from math import sqrt

mse = mean_squared_error(y_test,y_pred)
rmse = sqrt(mse)
rmse

4.453237437198149

In [17]:
#r squared train data
model.score(x_train,y_train)

0.7103879080674731

In [18]:
#margin of error: RMSE

y_pred_train = model.predict(x_train)

from sklearn.metrics import mean_squared_error
from math import sqrt

mse = mean_squared_error(y_train,y_pred_train)
rmse = sqrt(mse)
rmse

4.849055005805464

In [19]:
import statsmodels.formula.api as smf

reg_all = smf.ols(formula = 'MEDV ~ CRIM + ZN + CHAS + AGE + INDUS + NX + RM \
+ DIS + RAD + TAX + PTRATIO + B + LSTAT', data=boston).fit()
reg_all.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     108.1
Date:                Tue, 08 Aug 2023   Prob (F-statistic):          6.72e-135
Time:                        10:02:17   Log-Likelihood:                -1498.8
No. Observations:                 506   AIC:                             3026.
Df Residuals:                     492   BIC:                             3085.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     36.4595      5.103      7.144      0.000      26.432      46.487
CRIM          -0.1080      0.033     -3.287      0.001      -0.173      -0.043
ZN             0.0464      0.014      3.382      0.001       0.019       0.073
CHAS           2.6867      0.862      3.118      0.002       0.994       4.380
AGE            0.0007      0.013      0.052      0.958      -0.025       0.027
INDUS          0.0206      0.061      0.334      0.738      -0.100       0.141
NX           -17.7666      3.820     -4.651      0.000     -25.272     -10.262
RM             3.8099      0.418      9.116      0.000       2.989       4.631
DIS           -1.4756      0.199     -7.398      0.000      -1.867      -1.084
RAD            0.3060      0.066      4.613      0.000       0.176       0.436
TAX           -0.0123      0.004     -3.280      0.001      -0.020      -0.005
PTRATIO       -0.9527      0.131     -7.283      0.000      -1.210      -0.696
B              0.0093      0.003      3.467      0.001       0.004       0.015
LSTAT         -0.5248      0.051    -10.347      0.000      -0.624      -0.425
==============================================================================
Omnibus:                      178.041   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              783.126
Skew:                           1.521   Prob(JB):                    8.84e-171
Kurtosis:                       8.281   Cond. No.                     1.51e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.51e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [20]:
reg_all = smf.ols(formula = 'MEDV ~ CRIM + ZN + CHAS + NX + RM \
+ DIS + RAD + TAX + PTRATIO + B + LSTAT', data=boston).fit()
reg_all.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   MEDV   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.735
Method:                 Least Squares   F-statistic:                     128.2
Date:                Tue, 08 Aug 2023   Prob (F-statistic):          5.54e-137
Time:                        10:02:37   Log-Likelihood:                -1498.9
No. Observations:                 506   AIC:                             3022.
Df Residuals:                     494   BIC:                             3072.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     36.3411      5.067      7.171      0.000      26.385      46.298
CRIM          -0.1084      0.033     -3.307      0.001      -0.173      -0.044
ZN             0.0458      0.014      3.390      0.001       0.019       0.072
CHAS           2.7187      0.854      3.183      0.002       1.040       4.397
NX           -17.3760      3.535     -4.915      0.000     -24.322     -10.430
RM             3.8016      0.406      9.356      0.000       3.003       4.600
DIS           -1.4927      0.186     -8.037      0.000      -1.858      -1.128
RAD            0.2996      0.063      4.726      0.000       0.175       0.424
TAX           -0.0118      0.003     -3.493      0.001      -0.018      -0.005
PTRATIO       -0.9465      0.129     -7.334      0.000      -1.200      -0.693
B              0.0093      0.003      3.475      0.001       0.004       0.015
LSTAT         -0.5226      0.047    -11.019      0.000      -0.616      -0.429
==============================================================================
Omnibus:                      178.430   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              787.785
Skew:                           1.523   Prob(JB):                    8.60e-172
Kurtosis:                       8.300   Cond. No.                     1.47e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.47e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**RIDGE REGRESSION**

Without cross validation and tuning

In [21]:
import pandas as pd

boston = pd.read_csv("boston.csv")

boston.head()

CRIM    ZN  INDUS  CHAS     NX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

In [22]:
x = boston.drop("MEDV", axis=1).values #input factors

y = boston["MEDV"].values
#values method converts a datafarme into array

In [23]:
from sklearn.model_selection import RepeatedKFold
from sklearn.linear_model import Ridge
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.3,random_state=1)

#instantiate the model
model = Ridge(alpha=1)

model.fit(x_train,y_train)


Ridge(alpha=1)

In [24]:
#r squared
model.score(x_test,y_test)

0.7890510666829776

In [5]:
y_pred = model.predict(x_test)

#margin of error
from sklearn.metrics import mean_squared_error

mse = mean_squared_error(y_test, y_pred)
rmse = mse ** 0.5
rmse

4.397091798887489

In [25]:
#overfitting
y_pred_train = model.predict(x_train)

#margin of error
from sklearn.metrics import mean_squared_error

mse = mean_squared_error(y_train, y_pred_train)
rmse = mse ** 0.5
rmse

4.882557479143369

In [26]:
model.intercept_

38.95161909934214

In [27]:
model.coef_

array([-8.99352520e-02,  6.20345865e-02,  1.21404325e-02,  2.23426149e+00,
       -1.12838152e+01,  2.89618901e+00, -4.81458007e-03, -1.36998976e+00,
        2.83653073e-01, -1.22828776e-02, -8.84229846e-01,  7.09753443e-03,
       -5.80033848e-01])

In [28]:
o = boston.drop("MEDV", axis=1)
o.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT'],
      dtype='object')

In [29]:
pd.DataFrame(model.coef_, index=o.columns)

0
CRIM     -0.089935
ZN        0.062035
INDUS     0.012140
CHAS      2.234261
NX      -11.283815
RM        2.896189
AGE      -0.004815
DIS      -1.369990
RAD       0.283653
TAX      -0.012283
PTRATIO  -0.884230
B         0.007098
LSTAT    -0.580034

**Ridge Regression**

With crossvalidation and tuning

RidgeCV: gridsearch is inbuilt

In [30]:
from sklearn.linear_model import RidgeCV
from numpy import arange

#instantiate repeated k fold
rcv = RepeatedKFold(n_splits=10,n_repeats=3,random_state=1)

#instantiate ridgecv
model = RidgeCV(alphas=arange(0.1, 1, 0.01), cv=rcv, scoring='neg_mean_absolute_error')

model.fit(x_train, y_train)

RidgeCV(alphas=array([0.1 , 0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 ,
       0.21, 0.22, 0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 , 0.31,
       0.32, 0.33, 0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 , 0.41, 0.42,
       0.43, 0.44, 0.45, 0.46, 0.47, 0.48, 0.49, 0.5 , 0.51, 0.52, 0.53,
       0.54, 0.55, 0.56, 0.57, 0.58, 0.59, 0.6 , 0.61, 0.62, 0.63, 0.64,
       0.65, 0.66, 0.67, 0.68, 0.69, 0.7 , 0.71, 0.72, 0.73, 0.74, 0.75,
       0.76, 0.77, 0.78, 0.79, 0.8 , 0.81, 0.82, 0.83, 0.84, 0.85, 0.86,
       0.87, 0.88, 0.89, 0.9 , 0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97,
       0.98, 0.99]),
        cv=RepeatedKFold(n_repeats=3, n_splits=10, random_state=1),
        scoring='neg_mean_absolute_error')

In [31]:
# summarize chosen configuration
model.alpha_

0.5799999999999997

In [32]:
#rsquared
model.score(x_test,y_test)

0.7883167600760095

In [33]:
y_pred = model.predict(x_test)

mse = mean_squared_error(y_test,y_pred)
rmse = mse ** 0.5
rmse

4.4047382196086

In [34]:
#overfitting
y_pred_train = model.predict(x_train)

mse = mean_squared_error(y_train,y_pred_train)
rmse = mse ** 0.5
rmse

4.866642733557797

Linear regression

R = 78.36, rmse = 4.45

Ridge without cv and tuning

R = 78.9, rmse=4.40

Ridge with cv and tuning

R = 78.83, rmse = 4.4047

In [35]:
model.coef_

array([-9.22605745e-02,  6.16667500e-02,  2.50763986e-02,  2.30126902e+00,
       -1.40928582e+01,  2.87272384e+00, -2.54431729e-03, -1.41037774e+00,
        2.90013469e-01, -1.19966062e-02, -9.17381421e-01,  6.92039959e-03,
       -5.76631711e-01])

In [36]:
pd.DataFrame(model.coef_, index=o.columns)

0
CRIM     -0.092261
ZN        0.061667
INDUS     0.025076
CHAS      2.301269
NX      -14.092858
RM        2.872724
AGE      -0.002544
DIS      -1.410378
RAD       0.290013
TAX      -0.011997
PTRATIO  -0.917381
B         0.006920
LSTAT    -0.576632

**Lasso Regression**

without cv and tuning

ridge (L2) = SSR + lambda * slope square

lasso (L1) = SSR + lambda * |slope|

In [37]:
from sklearn.linear_model import Lasso

#instantiate lasso
model = Lasso(alpha=1)

model.fit(x_train,y_train)

Lasso(alpha=1)

In [38]:
#r squared
model.score(x_test,y_test)

0.6694782854622285

In [39]:
y_pred = model.predict(x_test)

mse = mean_squared_error(y_test,y_pred)
rmse = mse ** 0.5
rmse

5.503980216349544

**Lasso with cross validation and tuning**

LassoCV

In [40]:
from sklearn.linear_model import LassoCV

#instantiate repeated k fold
rcv = RepeatedKFold(n_splits=10,n_repeats=3,random_state=1)

model = LassoCV(alphas=arange(0,1,0.01),\
                cv=rcv)#default scoring is r squared

model.fit(x_train,y_train)

C:\Users\Yashika Batra\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:633: UserWarning: Coordinate descent with alpha=0 may lead to unexpected results and is discouraged.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Yashika Batra\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:633: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 3825.9654070020806, tolerance: 2.481023952830189
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Yashika Batra\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:633: UserWarning: Coordinate descent with alpha=0 may lead to unexpected results and is discouraged.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Yashika Batra\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:633: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. 

C:\Users\Yashika Batra\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:633: UserWarning: Coordinate descent with alpha=0 may lead to unexpected results and is discouraged.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Yashika Batra\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:633: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 3931.0372351075566, tolerance: 2.7137653354231976
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Yashika Batra\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:633: UserWarning: Coordinate descent with alpha=0 may lead to unexpected results and is discouraged.
  model = cd_fast.enet_coordinate_descent_gram(
C:\Users\Yashika Batra\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:633: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations.

LassoCV(alphas=array([0.  , 0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 ,
       0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 , 0.21,
       0.22, 0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 , 0.31, 0.32,
       0.33, 0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 , 0.41, 0.42, 0.43,
       0.44, 0.45, 0.46, 0.47, 0.48, 0.49, 0.5 , 0.51, 0.52, 0.53, 0.54,
       0.55, 0.56, 0.57, 0.58, 0.59, 0.6 , 0.61, 0.62, 0.63, 0.64, 0.65,
       0.66, 0.67, 0.68, 0.69, 0.7 , 0.71, 0.72, 0.73, 0.74, 0.75, 0.76,
       0.77, 0.78, 0.79, 0.8 , 0.81, 0.82, 0.83, 0.84, 0.85, 0.86, 0.87,
       0.88, 0.89, 0.9 , 0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98,
       0.99]),
        cv=RepeatedKFold(n_repeats=3, n_splits=10, random_state=1))

In [41]:
model.alpha_

0.0

alpha 1 means model is same as traditional linear regression

In [42]:
model.score(x_test,y_test)

0.7836295385076293

In [43]:
y_pred = model.predict(x_test)

mse = mean_squared_error(y_test,y_pred)
rmse = mse ** 0.5
rmse

4.453237437198148

In [44]:
pd.DataFrame(model.coef_,index=o.columns)

0
CRIM     -0.098542
ZN        0.060784
INDUS     0.059172
CHAS      2.439560
NX      -21.469965
RM        2.795814
AGE       0.003575
DIS      -1.516272
RAD       0.307542
TAX      -0.011280
PTRATIO  -1.005466
B         0.006450
LSTAT    -0.568835

Ridge with cv and tuning

R = 78.83, rmse=4.4047

Lasso with cv and tuning

R= 77.9, rmse = 4.49